src/util-napatech.c

```diff
@@ -23,28 +22,201 @@
+ *
+ * \return count of the Napatech adapters present in the system.
+ */
+static int GetNumAdapters(void)
+{
+    NtInfoStream_t hInfo;
+    NtInfo_t hInfoSys;
+    int status;
+
+    if ((status = NT_InfoOpen(&hInfo, "InfoStream")) != NT_SUCCESS) {
+        NAPATECH_ERROR(SC_ERR_NAPATECH_OPEN_FAILED, status);
+        exit(EXIT_FAILURE);
+    }
+
+    hInfoSys.cmd = NT_INFO_CMD_READ_SYSTEM;
+    if ((status = NT_InfoRead(hInfo, &hInfoSys)) != NT_SUCCESS) {
+        NAPATECH_ERROR(SC_ERR_NAPATECH_OPEN_FAILED, status);
+       exit(EXIT_FAILURE);
+    }
+
+    int num_adapters = hInfoSys.u.system.data.numAdapters;
+
+    NT_InfoClose(hInfo);
+    return num_adapters;
+}
+
+/**
+ * \brief  Initializes the FlowStreams used to program flow data.
+ *
+ * Opens a FlowStream on each adapter present in the system.  This
+ * FlowStream is subsequently used to program the adapter with
+ * flows to bypass.
+ *
+ * \return count of the Napatech adapters present in the system.
+ */
+int NapatechInitFlowStreams(void)
+{
+    int status;
+    int adapter = 0;
+    int num_adapters = GetNumAdapters();
+    SCLogInfo("Found %d Napatech adapters.\n", num_adapters);
+    memset(&hFlowStream, 0, sizeof(hFlowStream));
+
+    if (!NapatechUseHWBypass()) {
+        /* HW Bypass is disabled in the conf file */
+        return 0;
+    }
+
+    for (adapter = 0; adapter < num_adapters; ++adapter) {
+        NtFlowAttr_t attr;
+        char flow_name[80];
+
+        NT_FlowOpenAttrInit(&attr);
+        NT_FlowOpenAttrSetAdapterNo(&attr, adapter);
+
+        snprintf(flow_name, sizeof(flow_name), "Flow stream %d", adapter );
+        SCLogInfo("Opening flow programming stream:  %s\n", flow_name);
+        if ((status = NT_FlowOpen_Attr(&hFlowStream[adapter], flow_name, &attr)) != NT_SUCCESS) {
+            SCLogWarning(SC_WARN_COMPATIBILITY, "Napatech bypass functionality not supported by the FPGA version on adapter %d - disabling support.", adapter);
+            bypass_supported = 0;
+            return 0;
+        }
+    }
+
+    bypass_supported = 1;
+    return bypass_supported;
+}
+
+/**
+ * \brief  Returns a pointer to the FlowStream associated with this adapter.
+ *
+ * \return count of the Napatech adapters present in the system.
+ */
+NtFlowStream_t *NapatechGetFlowStreamPtr(int device)
+{
+    return &hFlowStream[device];
+}
+
+/**
+ * \brief Closes all open FlowStreams
+ *
+ * \return Success of the operation.
+ */
+int NapatechCloseFlowStreams(void)
+{
+    int status = 0;
+    int adapter = 0;
+    int num_adapters = GetNumAdapters();
+
+    for (adapter = 0; adapter < num_adapters; ++adapter) {
+        if (hFlowStream[adapter]) {
+            SCLogInfo("Closing Napatech Flow Stream on adapter %d.", adapter);
+            if ((status = NT_FlowClose(hFlowStream[adapter])) != NT_SUCCESS) {
+                NAPATECH_ERROR(SC_ERR_SHUTDOWN, status);
+            }
+            hFlowStream[adapter] = NULL;
+        }
+    }
+    return (status == NT_SUCCESS);
```

Comment:
Only the status from the last close is returned; if an error occurs previously, it'll be lost.